### PR TITLE
Feature/#85

### PR DESCRIPTION
**Pull request by remi128**
implementation of issue #85 data versioning and conversion

### CODE REVIEW

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/AfterUpdateHandler.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ * #L%
+ */
+package de.braintags.vertx.jomnigate.dataaccess.write.impl;
+
+import de.braintags.vertx.jomnigate.observer.ObserverEventType;
+
+/**
+ * 
+ *
```
Comment (sschmittBt) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/AfterUpdateHandler.java]: Missing comment

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeInsertHandler.java [hunk 2]
```diff
@@ -13,49 +13,71 @@
...
+  protected <T> CompositeFuture loopObserver(List<IObserver> ol, IWrite<T> writeObject, T entity,
       IObserverContext context) {
     List<Future> fl = new ArrayList<>();
-    Iterator<?> selection = ((AbstractWrite) writeObject).getSelection();
-    while (selection.hasNext()) {
-      IObserverEvent event = IObserverEvent.createEvent(ObserverEventType.BEFORE_SAVE, selection.next(), null,
-          writeObject, writeObject.getDataStore());
+    for (IObserver observer : ol) {
+      IObserverEvent event = IObserverEvent.createEvent(getEventType(), entity, null, writeObject,
+          writeObject.getDataStore());
```
Comment (sschmittBt) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeInsertHandler.java]: Does every observer need its own event, or can you move the event creation outside of the loop?

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeUpdateHandler.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ * #L%
+ */
+package de.braintags.vertx.jomnigate.dataaccess.write.impl;
+
+import de.braintags.vertx.jomnigate.observer.ObserverEventType;
+
+/**
+ *
```
Comment (sschmittBt) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeUpdateHandler.java]: Missing comment

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/annotation/VersionInfo.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE })
+public @interface VersionInfo {
+
+  /**
+   * Defines the event type, during which defined converters shall be executed
+   * 
+   * @return
+   */
+  ObserverEventType eventType() default ObserverEventType.BEFORE_UPDATE;
```
Comment (sschmittBt) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/annotation/VersionInfo.java]: Could you add the allowed event types to the comment?

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapper.java [hunk 5]
```diff
@@ -175,38 +201,52 @@ protected void computeEntity() {
...
       }
     }
+    computeIndize(definitions);
+    this.indexes = ImmutableSet.copyOf(definitions.values());
+  }
 
+  /**
+   * @param definitions
+   */
+  private void computeIndize(Map<String, IIndexDefinition> definitions) {
```
Comment (sschmittBt) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapper.java]: Either indexes or indices, but not indize

**Review by sschmittBt — CHANGES_REQUESTED** [review]
Instead of the IMapperVersion interface, that every class has to implement if it wants versioning, wouldn't it be nicer if the version info would be added dynamically to saved entities? Similar to how Jackson handles the @class values for polymorphism. This way, the user doesn't need to care for anything, can not manipulate the version and produce unexpected behaviour, and can simply control everything via the annotation.

File: vertx-pojo-mapper-common-test/src/main/java/de/braintags/vertx/jomnigate/testdatastore/TestDataVersioning.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    VersioningWithInterface_V5 vi = new VersioningWithInterface_V5();
+    saveRecord(context, vi);
+    context.assertEquals(5l, vi.getMapperVersion(), "version was not automatically set");
+    VersioningWithInterface_V5 vi2 = findRecordByID(context, VersioningWithInterface_V5.class, vi.id);
+    context.assertEquals(5l, vi2.getMapperVersion(), "version was NOT saved");
+
+    // set the version manually, save the record, it should NOT be updated
+    vi2.setMapperVersion(88);
+    saveRecord(context, vi2);
+    context.assertEquals(88l, vi2.getMapperVersion(), "for an update the version should NOT be set");
```
Comment (sschmittBt) [inline vertx-pojo-mapper-common-test/src/main/java/de/braintags/vertx/jomnigate/testdatastore/TestDataVersioning.java:120]: Why should the new version be "881" if it was "51" before and you set it to "88"?

**Comment by remi128** [comment]
yes, would be nice, especially cause user would not be able to manipulate the version number. Unfortunately i need the version number in access from an entity, when the versioning is running in phase BEFORE_UPDATE for instance.

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/AfterUpdateHandler.java (hunk 1 above)
Comment (remi128) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/AfterUpdateHandler.java]: done

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeInsertHandler.java (hunk 2 above)
Comment (remi128) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeInsertHandler.java]: good - done

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeUpdateHandler.java (hunk 3 above)
Comment (remi128) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/write/impl/BeforeUpdateHandler.java]: done

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/annotation/VersionInfo.java (hunk 4 above)
Comment (remi128) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/annotation/VersionInfo.java]: done

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapper.java (hunk 5 above)
Comment (remi128) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapper.java]: done

**Review by sschmittBt — CHANGES_REQUESTED** [review]

File: vertx-pojo-mapper-common-test/src/main/java/de/braintags/vertx/jomnigate/testdatastore/TestDataVersioning.java (hunk 6 above)
Comment (remi128) [reply vertx-pojo-mapper-common-test/src/main/java/de/braintags/vertx/jomnigate/testdatastore/TestDataVersioning.java:120]: 88l -> 88 Long

Comment (sschmittBt) [reply vertx-pojo-mapper-common-test/src/main/java/de/braintags/vertx/jomnigate/testdatastore/TestDataVersioning.java:120]: Wow... that "l" looks seriously like a 1 in the comparison... my bad.

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapper.java (hunk 5 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapper.java]: Did you commit it? For me it still says "indize"

Comment (remi128) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapper.java]: ups - second method, pushed now

**Review by sschmittBt — APPROVED** [review]

**Pull request merged**